system/jlib/jstats.cpp

```diff
@@ -1783,11 +1786,78 @@ class SortedCollectionIterator : public ArrayIIteratorOf<IArrayOf<IStatisticColl
 class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
 {
     friend class CollectionHashTable;
+
+    CStatisticCollection * ensureSubScopePath(std::initializer_list<const StatsScopeId> path)
+    {
+        CStatisticCollection * curScope = this;
+        for (const auto & scopeItem: path)
+            curScope = curScope->ensureSubScope(scopeItem, true); // n.b. this will always return a valid pointer
+        return curScope;
+    }
+    void markDirty()
+    {
+        isDirty=true;
+        if (parent) parent->markDirty();
+    }
+    void refreshAggregates(CRuntimeStatisticCollection & parentTotals, std::function<void(const char * scope, StatisticScopeType sst, StatisticKind kind, stat_type value)> & fWhenAggregateUpdated)
+    {
+        const StatisticsMapping & mapping = parentTotals.queryMapping();
+        // if this scope is not dirty, the aggregates are accurate at this level so return totals (no need to descend)
+        // Also there is no stats below sg scope, so no need to descend
+        if (isDirty==false || id.queryScopeType()==SSTsubgraph)
+        {
+            // return the aggregates at this scope level
+            ForEachItemIn(i, stats)
+            {
+                Statistic & stat = stats.element(i);
+                StatisticKind kind = stat.queryKind();
+                if (kind != (StatisticKind)(kind & StKindMask))
```

Comment:
Easier to read as 

    queryStatsVariant(kind) != 0

(mentioned elsewhere).  Possibly better would be to move that test into hasKind()